lib/df-detection.c

```diff
@@ -0,0 +1,133 @@
+#include <linux/entry-common.h>
+#include <linux/kernel.h>
+#include <linux/livepatch.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+void add_address(const void *addr, size_t len, unsigned long caller)
+{
+	if (current->addresses == NULL || current->pairs == NULL)
+		return;
+	if (current->num_read >= current->sz &&
+	    reallocate_extra_memory(current->sz, DF_MAX_RECORDS)) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		detect_intersection();
+		current->num_read++;
+	}
+}
+
+void start_system_call(long syscall)
+{
+	if (syscall == 54 || syscall == 165 || syscall == 16)
+		return;
+	current->syscall_num = syscall;
+	current->addresses = (struct df_address_range *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
+	current->sz = current->addresses ? DF_INIT_SIZE : 0;
+	current->num_read = 0;
+	current->pairs = (struct df_pair *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
+	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+	current->df_index = 0;
+}
+void end_system_call(void)
+{
+	if (current->pairs != NULL) {
+		if (current->df_index)
+			report();
+		kfree(current->pairs);
+		current->pairs = NULL;
+		current->df_index = 0;
+		current->df_size = 0;
+	}
+	if (current->addresses != NULL) {
+		current->num_read = 0;
+		current->sz = 0;
+		kfree(current->addresses);
+		current->addresses = NULL;
+	}
+}
+void report(void)
+{
+	int i;
+	pr_err("BUG: Intersection Detected at syscall: %pSR\n ",
+	       sys_call_table[current->syscall_num]);
+	pr_err("==================================================================\n");
+	pr_err("syscall number %ld  System Call: %pSR\n", current->syscall_num,
+	       sys_call_table[current->syscall_num]);
+	for (i = 0; i < current->df_index; i++) {
+		pr_err("First %px len %lu Caller %pSR \nSecond %px len "
+		       "%lu Caller %pSR \n",
+		       current->pairs[i].first->start_address,
+		       current->pairs[i].first->len,
+		       current->pairs[i].first->caller,
+		       current->pairs[i].second->start_address,
+		       current->pairs[i].second->len,
+		       current->pairs[i].second->caller);
+	}
+	pr_err("==================================================================\n");
+	if (panic_on_warn) {
+		panic("panic_on_warn set. \n");
+		panic_on_warn = 0;
+	}
+}
+
+// it returns 0 when it fails to re allocate memory
+int reallocate_extra_memory(int sz, int mx_size)
```

Comment:
I think you don't need this function anymore.

Reply:
in syzkaller detected that this warning happened 4 times

Reply:
True; what I was trying to say is that the function doesn't really match its name anymore, it just checks the size instead of actually reallocating memory.
You could either perform the check in-place instead of calling this function (it's basically equivalent to `!WARN_ON(sz, mx_size)`, or pull krealloc into this function as well, so that it encapsulates memory reallocation.